// @astrouxds/astro-web-components/dist/components/rux-menu-item.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const ruxMenuItemCss = ":host{display:block}::slotted(*:focus-visible){outline:2px solid transparent}a.rux-menu-item,a.rux-menu-item:hover{text-decoration:none}.rux-menu-item{display:block;position:relative;padding:var(--spacing-2, 0.5rem);color:var(--color-text-interactive-default, #4dacff);font-family:var(--font-control-body-1-font-family, 'Roboto', -apple-system, BlinkMacSystemFont, 'Segoe UI', Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif);font-size:var(--font-control-body-1-font-size, 1rem);font-weight:var(--font-control-body-1-font-weight, 400);line-height:var(--font-control-body-1-line-height, calc(20 / 16));letter-spacing:var(--font-control-body-1-letter-spacing, 0.005em);cursor:pointer}.rux-menu-item:hover{color:var(--color-text-interactive-hover, #92cbff);background:var(--color-background-base-hover, #142435)}.rux-menu-item:focus-visible{border-radius:var(--radius-base, 3px);outline:var(--border-width-focus-default, 1px) solid var(--color-border-focus-default, #da9ce7);outline-offset:var(--spacing-focus-default, 0.125rem)}.rux-menu-item--selected{background:var(--color-background-base-selected, #1c3f5e);color:var(--color-text-primary, #ffffff)}.rux-menu-item--disabled{opacity:var(--opacity-disabled, 40%);cursor:not-allowed}.rux-menu-item--disabled:hover{background:var(--color-background-base-default, #101923);color:var(--color-text-interactive-default, #4dacff)}";

const RuxMenuItem$1 = /*@__PURE__*/ proxyCustomElement(class RuxMenuItem extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.selected = false;
    this.disabled = false;
    this.value = undefined;
    this.href = undefined;
    this.target = undefined;
    this.rel = undefined;
    this.download = undefined;
  }
  render() {
    const { disabled, selected, href, rel, download, target } = this;
    const TagType = href ? 'a' : 'div';
    const attributes = TagType === 'a' ? { download, href, rel, target } : {};
    return (h(Host, { value: this.value }, h(TagType, Object.assign({}, attributes, { class: {
        'rux-menu-item': true,
        'rux-menu-item--selected': selected,
        'rux-menu-item--disabled': disabled,
      } }), h("slot", null))));
  }
  get el() { return this; }
  static get style() { return ruxMenuItemCss; }
}, [1, "rux-menu-item", {
    "selected": [516],
    "disabled": [516],
    "value": [1025],
    "href": [1],
    "target": [1],
    "rel": [1],
    "download": [1]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["rux-menu-item"];
  components.forEach(tagName => { switch (tagName) {
    case "rux-menu-item":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, RuxMenuItem$1);
      }
      break;
  } });
}

const RuxMenuItem = RuxMenuItem$1;
const defineCustomElement = defineCustomElement$1;

export { RuxMenuItem, defineCustomElement };
