// @astrouxds/astro-web-components/dist/collection/components/rux-timeline/rux-ruler/rux-ruler.js
import { Host, h } from '@stencil/core';
import { dateRange as getRange } from '../helpers';
export class RuxRuler {
  constructor() {
    this.timePattern = /^00:.+$/;
    this.interval = '';
    this.start = '';
    this.end = '';
    this.timezone = 'UTC';
    this.showStartOfDay = false;
  }
  get dateRange() {
    return getRange(new Date(this.start), new Date(this.end), this.interval, 1, this.timezone);
  }
  getColumn(index) {
    let unitOfTime = 60;
    if (this.interval === 'day') {
      unitOfTime = 24;
    }
    // same as for days
    if (this.interval === 'week') {
      unitOfTime = 24;
    }
    if (this.interval === 'month') {
      unitOfTime = 24;
    }
    const start = unitOfTime * index + 2;
    const end = start + unitOfTime;
    return `${unitOfTime * index + 2} / ${end}`;
  }
  shouldShowDate(time) {
    if (this.interval !== 'hour') {
      return false;
    }
    if (!this.showStartOfDay) {
      return false;
    }
    return this.timePattern.test(time);
  }
  render() {
    return (h(Host, null, h("div", { class: "rux-ruler rux-track" }, this.dateRange.map(([time, newDayDate], index) => {
      const newDay = this.timePattern.test(time)
        ? newDayDate
        : '';
      return (h("span", { key: index, class: {
          'ruler-time': true,
          'ruler-new-day-cell': this.shouldShowDate(time),
        }, style: {
          gridRow: '1',
          gridColumn: this.getColumn(index),
        } }, time, this.shouldShowDate(time) ? (h("span", { class: "ruler-new-day-display" }, newDay)) : ('')));
    }))));
  }
  static get is() { return "rux-ruler"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["rux-ruler.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["rux-ruler.css"]
    };
  }
  static get properties() {
    return {
      "interval": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "The Timeline's interval. Set automatically from the parent Timeline component"
            }],
          "text": ""
        },
        "attribute": "interval",
        "reflect": false,
        "defaultValue": "''"
      },
      "start": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "The Timeline's start date. Set automatically from the parent Timeline component"
            }],
          "text": ""
        },
        "attribute": "start",
        "reflect": false,
        "defaultValue": "''"
      },
      "end": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "The Timeline's end date. Set automatically from the parent Timeline component"
            }],
          "text": ""
        },
        "attribute": "end",
        "reflect": false,
        "defaultValue": "''"
      },
      "timezone": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "- The Ruler's time zone. Set automatically from the parent Timeline component."
            }],
          "text": ""
        },
        "attribute": "timezone",
        "reflect": true,
        "defaultValue": "'UTC'"
      },
      "showStartOfDay": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean | undefined",
          "resolved": "boolean | undefined",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [],
          "text": "Display the day (MM/DD) at 00:00. Only works when Timeline interval is set to 'hour'."
        },
        "attribute": "show-start-of-day",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get elementRef() { return "el"; }
}
