// @astrouxds/astro-web-components/dist/tests/utils/_astro-fixtures.js
// import { test as base } from '@playwright/test'
import { test as base } from "@astrouxds/stencil-playwright";
export const test = base.extend({
  theme: ['dark', { option: true }],
  component: ['rux-foo', { option: true }],
  testName: ['basic', { option: true }],
  astroVRTPage: async ({ theme, component, testName, page }, use) => {
    await page.goto(`/src/components/${component}/test/${testName}/index.html`);
    // Set up the fixture.
    if (theme === 'light') {
      await page.evaluate(() => {
        document.body.classList.add('light-theme');
      });
    }
    // Use the fixture value in the test.
    await use(page);
  },
  // astroPage: async ({ page }, use) => {
  //     // Set up the fixture.
  //     const astroPage = new AstroComponentPage(page)
  //     await astroPage.goto()
  //     // Use the fixture value in the test.
  //     await use(astroPage)
  //     // Clean up the fixture.
  //     //   await todoPage.removeAll();
  // },
  // astroFormPage: async ({ page }, use) => {
  //     // Set up the fixture.
  //     const astroPage = new AstroFormPage(page)
  //     await astroPage.goto()
  //     // Use the fixture value in the test.
  //     await use(astroPage)
  //     // Clean up the fixture.
  //     //   await todoPage.removeAll();
  // }
});
export { expect } from '@playwright/test';
